### ISSUE TITLE
[SOLVED] TT-RSS doesn't connect. 

### ISSUE THREAD

**Issue by PolGZ** [Suzie97/Communique#11]
## Describe the bug
Login to Tiny Tiny RSS doesn't seem to work, and doesn't give any error message back.

## To Reproduce

1. Use my (already tested) address to my tt-rss server, user and stored password
2. Hit login
3. Nothing happens

## Expected behavior
Log in successfully (or at least give an error message explaining why it doesn't work)

## Platform Information
Flatpak version 1.1.0

---

Should any API plugin be enabled at server side to run with communique?

**Comment by Suzie97** [comment]
Hey @PolGZ, sorry for the inconvenience. You can follow the instructions here to make it work: https://github.com/Suzie97/Communique/blob/master/data/tt-rss-feedreader-plugin/README.md

I'll try to build this in the app itself and make it work out of the box.

**Comment by PolGZ** [comment]
> Hey @PolGZ, sorry for the inconvenience. You can follow the instructions here to make it work: https://github.com/Suzie97/Communique/blob/master/data/tt-rss-feedreader-plugin/README.md
> 
Hi, thanks for the reply.

I don't know why it didn't say anything when I tried on my laptop, and I actually wondered if it would be something like that, as FeedReader was my go to reader. I then got the error message on my other computer and now I see it here as well.

The thing is, I reinstalled TT-RSS now with the "official" docker way (since my raspberry SSD died and I decided it was a good time to migrate to 64bit OS). I actually enabled the api_feedreader on TT-RSS, and it shows as activated on settings, but both feedreader and communique keep saying "install the api_feedreader plugin on your tt-rss instance". In fact, the reason I found communique was searching for another tt-rss desktop reader, since feedreader gave me that error.

Are you actually able to run it with a docker install of tt-rss?

**Comment by Suzie97** [comment]
@PolGZ can you mention the feedreader version after which ttrss stopped working for you?

**Comment by PolGZ** [comment]
> @PolGZ can you mention the feedreader version after which ttrss stopped working for you?

Honestly, I couldn't tell. I installed the current version from the [docker repo](https://git.tt-rss.org/fox/ttrss-docker-compose). On the web interface it says `v21.12-7aeaa1b03`. It's the first time I use docker, so I don't know if there is a way to check the version of the tt-rss server itself.

Before that, I never actually updated the version I originally installed, since FeedReader stopped developing and I was afraid to loose compatibility. I know I installed it around 07/02/2021 (that's the date of creation of my installation notes, and I always write them down as I progress within a given installation), pulling from git the current version by that time.

**Comment by PolGZ** [comment]
> @PolGZ can you mention the feedreader version after which ttrss stopped working for you?

I misread your question, @Suzie97. I thought about the tt-rss version. But the FeedReader version I installed back in the day was already the last one [(v2.11.0)](https://github.com/jangernert/FeedReader/releases/tag/v2.11.0). I started using it a few months after that last release.

I asked in the [tt-rss forum (section about plugins)](https://community.tt-rss.org/t/feedreader-plugin-still-working-tt-rss-docker/5143) if anybody used the feedreader plugin, but it doesn't seem so. They told me, though, about [recent changes that required some plugins to be updated](https://community.tt-rss.org/t/heads-up-ongoing-attempt-to-comply-with-phpstan-rule-level-6/5046). I don't really understand the scope of those changes, but maybe you can understand it.

Thanks for the attention, man.

**Comment by Suzie97** [comment]
> > @PolGZ can you mention the feedreader version after which ttrss stopped working for you?
> 
> I misread your question, @Suzie97. I thought about the tt-rss version. But the FeedReader version I installed back in the day was already the last one [(v2.11.0)](https://github.com/jangernert/FeedReader/releases/tag/v2.11.0). I started using it a few months after that last release.
> 
> I asked in the [tt-rss forum (section about plugins)](https://community.tt-rss.org/t/feedreader-plugin-still-working-tt-rss-docker/5143) if anybody used the feedreader plugin, but it doesn't seem so. They told me, though, about [recent changes that required some plugins to be updated](https://community.tt-rss.org/t/heads-up-ongoing-attempt-to-comply-with-phpstan-rule-level-6/5046). I don't really understand the scope of those changes, but maybe you can understand it.
> 
> Thanks for the attention, man.

Thanks a lot for the info. This might be related as other plugins seem to have broken too. I'll look into this

**Comment by lihz1990** [comment]
@PolGZ  I just spend some time on the source code of ttrss and Communique. If you change code of `api_feedreader` plugin in ttrss like below, your feedreader/Communique will work fine, wiith removeLabel function broken.

```Php
……
function removeLabel()
{
	$label_id = (int)$_REQUEST["label_id"];
	if($label_id != "")
	{
		return array(API::STATUS_ERR, array("error" => 'INCORRECT_USAGE'));   # <= just add this line
		Labels::remove(Labels::feed_to_label_id($label_id), $_SESSION["uid"]);
		return array(API::STATUS_OK);
	}
	else
	{
		return array(API::STATUS_ERR, array("error" => 'INCORRECT_USAGE'));
	}
}
```

The problem  ralated to functions below
file `Communique/plugins/backend/ttrss/ttrssAPI.vala`, function `haveAPIplugin`
file `Communique/data/tt-rss-feedreader-plugin/api_feedreader/init.php`, function `removeLabel`

As I don't know how much about php and Vala, I'm not able to completely fix this bug, But glade to provide more details if needed.

**Comment by Suzie97** [comment]
> @PolGZ I just spend some time on the source code of ttrss and Communique. If you change code of `api_feedreader` plugin in ttrss like below, your feedreader/Communique will work fine, wiith removeLabel function broken.
> 
> ```
> ……
> function removeLabel()
> {
> 	$label_id = (int)$_REQUEST["label_id"];
> 	if($label_id != "")
> 	{
> 		return array(API::STATUS_ERR, array("error" => 'INCORRECT_USAGE'));   # <= just add this line
> 		Labels::remove(Labels::feed_to_label_id($label_id), $_SESSION["uid"]);
> 		return array(API::STATUS_OK);
> 	}
> 	else
> 	{
> 		return array(API::STATUS_ERR, array("error" => 'INCORRECT_USAGE'));
> 	}
> }
> ```
> 
> The problem ralated to functions below file `Communique/plugins/backend/ttrss/ttrssAPI.vala`, function `haveAPIplugin` file `Communique/data/tt-rss-feedreader-plugin/api_feedreader/init.php`, function `removeLabel`
> 
> As I don't know how much about php and Vala, I'm not able to completely fix this bug, But glade to provide more details if needed.

@lihz1990 Thank you so much for looking into the issue. Are there any other issues you have observed after implementing this fix? If things work fine using your fix, we could merge that, as something is better than nothing.

CC @PolGZ , if you manage to test this fix, let me know of the results.

**Comment by PolGZ** [comment]
> @PolGZ I just spend some time on the source code of ttrss and Communique. If you change code of `api_feedreader` plugin in ttrss like below, your feedreader/Communique will work fine, wiith removeLabel function broken.
> 
> ```
> ……
> function removeLabel()
> {
> 	$label_id = (int)$_REQUEST["label_id"];
> 	if($label_id != "")
> 	{
> 		return array(API::STATUS_ERR, array("error" => 'INCORRECT_USAGE'));   # <= just add this line
> 		Labels::remove(Labels::feed_to_label_id($label_id), $_SESSION["uid"]);
> 		return array(API::STATUS_OK);
> 	}
> 	else
> 	{
> 		return array(API::STATUS_ERR, array("error" => 'INCORRECT_USAGE'));
> 	}
> }
> ```
> 
> The problem ralated to functions below file `Communique/plugins/backend/ttrss/ttrssAPI.vala`, function `haveAPIplugin` file `Communique/data/tt-rss-feedreader-plugin/api_feedreader/init.php`, function `removeLabel`
> 
> As I don't know how much about php and Vala, I'm not able to completely fix this bug, But glade to provide more details if needed.

Thank you so much for your contribution.

> CC @PolGZ , if you manage to test this fix, let me know of the results.

I don't have any programming knowledge so I don't really understand how this works, but I'll try in a couple of hours when I get home. I'll update the post.

Thanks, guys!

**Comment by lihz1990** [comment]
@Suzie97  I have submit a completely  fix PR for this issue. Please review  it.

**Comment by PolGZ** [comment]
> @Suzie97 I have submit a completely fix PR for this issue. Please review it.

Using the fix submitted by @lihz1990 [pull request](https://github.com/Suzie97/Communique/pull/13), I have managed to successfully log into my TT-RSS server.

Just to make it easier for other users:

Once you download the [api_feedreader from the original FeedReader project](https://github.com/jangernert/FeedReader/releases), as explained by @Suzie97 [here](https://github.com/Suzie97/Communique/tree/master/data/tt-rss-feedreader-plugin), you will have to edit the only file within that folder (`init.php`).

Where it said:

```
function removeLabel()
	{
		$label_id = (int)$_REQUEST["label_id"];
		if($label_id != "")
```

That last line should be: (just changing quotation marks for a 0)

```
function removeLabel()
	{
		$label_id = (int)$_REQUEST["label_id"];
		if($label_id != 0)
```

And where does one copy that api_feedreader folder this days? now that the "official" installation for tt-rss is a docker container?

1. Check where the code of the tt-rss container is mounted:
    ```
    docker volume inspect ttrss-docker_app | grep Mountpoint
    ```
    In my case, it was: `/var/lib/docker/volumes/ttrss-docker_app/_data`
2. Within that folder, one should have a folder named `tt-rss/` and, within that one, a `plugins` folder (not the one named `plugins.local`, though)
3. Copy the api_feedreader folder with the modified `init.php` file.
4. As this is a system-enabled plugin for the server, not one would activate from withing the web interface (those are in `plugins.local` folder), you should edit the environment file of the docker (`.env`).
    1. Search for the line that starts with `TTRSS_PLUGINS=`
    2. Uncomment that line
    3. Add `api_feedreader` to it (in my case, the line looks like this):
        ```
        TTRSS_PLUGINS=auth_internal,note,nginx_xaccel,api_feedreader
        ```
5. Start the docker again
    ```
    docker-compose up --build -d
    ```

**Comment by Suzie97** [comment]
@PolGZ thank you so much for testing this feature and providing useful and detailed results. 

You have a great explanation of the steps there, I think we should merge @lihz1990's PR and update the instructions for the trss plugin. Also, since getting the tress-plugin to work required some amount of manual work, we would also need to update the marketing material for Communique, i.e. let users know that its not a straight forward process.

Let me know what you guys think about this.

Also, a ton of thanks to @lihz1990 for digging deep into the issue and finding a fix for it ☺️.

> Thanks to @Suzie97 for being so attentive maintainer. I'll "buy you a coffe" right now!
That really means a lot to me!! Thank you so much!! ❤️

**Comment by PolGZ** [comment]
I agree on changing instructions on the tt-rss section. Once you merge the changes, it shouldn't point to the original API on FeedReader's git, since it is no longer maintained and closed to changes. As it still works for the FeedReader app and the rest of the api is still the same... it should continue to be named api_feedreader, right?

In regards to the process of adding plug-ins to tt-rss in docker, you should point to [their wiki](https://tt-rss.org/wiki/Plugins), once the new API is merged. And maybe point to my previous comment?

By the way, @Suzie97, I get [this error (Stripe Card Updater - Payment section](https://stripe.com/en-ch/india-data-storage-faq) in buymeacoffee.com, as this impacts internationally issued cards. Do you want to send me another paying method? (XMR for example?)

**Comment by Suzie97** [comment]
> I agree on changing instructions on the tt-rss section. Once you merge the changes, it shouldn't point to the original API on FeedReader's git, since it is no longer maintained and closed to changes. But, since it still works for the original FeedReader app and the rest of the api is still the same... I think it should continue to be named api_feedreader, right?
> 
> In regards to the process of adding plug-ins to tt-rss in docker, you should point to [their wiki](https://tt-rss.org/wiki/Plugins), once the new API is merged. And maybe point to my previous comment if you think that would help. (If you finally merge the request and change the tt-rss instructions pointing to your own file, let me now, so I edit the comment and don't confuse users with the URL to the old one)

Cool, I'll merge the PR and close this issue. Updating the documentation should be handled on another issue.

> By the way, @Suzie97, I get [this error (Stripe Card Updater - Payment section)](https://stripe.com/en-ch/india-data-storage-faq) in buymeacoffee.com, as this impacts internationally issued cards. Do you want to send me another paying method? (XMR address for example?)

Oh, thanks for the info. I didn't know that. Unfortunately I can't send you an XMR address. You can try buying the app from the elementary OS AppCenter if that is convenient for you and see if that works. Also, my request for being approved on GitHub sponsors has been pending, so once that is verified, you can support through GitHub sponsors as well. Sorry for the inconvenience. Thanks a lot!

**Comment by PolGZ** [comment]
> Oh, thanks for the info. I didn't know that. Unfortunately I can't send you an XMR address. You can try buying the app from the elementary OS AppCenter if that is convenient for you and see if that works. Also, my request for being approved on GitHub sponsors has been pending, so once that is verified, you can support through GitHub sponsors as well. Sorry for the inconvenience. Thanks a lot!

Unfortunately I don't run Elementary OS. It doesn't seem to be possible from [the web](https://appcenter.elementary.io/com.github.suzie97.communique/). Let me know once you get verified (for real. Send me a DM over Mastodon. I'm usually in fosstodon.org community, same username all in lowercase)
